### ISSUE TITLE
[SIP plugin] ICE is started too early

### ISSUE THREAD

**Issue by saghul** [meetecho/janus-gateway#321]
I've been debugging a weird issue which in a nutshell consists of ICE failing for outgoing calls, even when it shouldn't.

After a lot of digging I narrowed it down to the fact that ICE is started too early. After [this commit](https://github.com/meetecho/janus-gateway/commit/cc9eb176) ICE will be started with the first trickled candidate after parsing the SDP. The problem is that we don't signal our candidates to the remote party until they answer the call, so the caller browser doesn't know where to send the connectivity checks, and Janus just times out and tears down the session.

The solution I found was to revert https://github.com/meetecho/janus-gateway/commit/cc9eb176, but the patch was committed for a reason, as it specifically addresses some problem, so I guess we need to find some way to solve it.

**Comment by saghul** [comment]
I forgot to mention how to reproduce it: just call from the SIP demo to a SIP address and don't answer, let it ring. It will fail in a few seconds.

**Comment by lminiero** [comment]
Probably related to #286. I couldn't replicate it, as discussed in a comment, but I only waited several seconds before answering (~10s) rather than just let it ring.

The commit you mentioned fixed a quite serious issue, as there were cases when ICE would never start at all. Not sure what the best way to proceed would be: this may be an issue in the browsers, as IIRC Chrome, for instance, has about a 30s timeout before giving up on ICE, which means that if it doesn't get an answer by then it gives up. We don't care in Janus instead: a ```failed``` should never be definitive, as with ```trickle``` it's easy to get in a ```failed``` state while you process candidates that arrive in bursts.

**Comment by saghul** [comment]
If browsers decide to fail, that's ok. The problem at hand is that Janus begins the ICE process without telling the browser: when the SDP from the browser is received, we convert it into a "boring" SDP and send it in a SIP INVITE. Then, while the other party is ringing we trickle the candidates, but as Janus acts as a controlled agent it won't begin connectivity checks, it will wait for the browser to start them, right? But the browser will never start them because it has no notion of the remote candidates.

What where those cases when ICE would never be started? AFAIS starting ICE unilaterally, without communicating our candidates to the browser so it can do RTCPeerConnection.addIceCandidate this is bound to fail due to a timeout.

Am I missing something?

**Comment by lminiero** [comment]
There were race conditions where you could receive all candidates (including the "end-of-candidates") _before_ receiving the offer from the browser: in this case, Janus had all the info it needed but never started ICE at all. I can't recall the specifics, but I do remember fighting with the bug for months (it was particularly evident when wrapping Janus requests on the server side, which could alter the original messaging) and my sigh of relief when I eventually fixed it for good.

Thinking about this, when I said Janus doesn't care about the ```failed``` state, this is only true while still receiving candidates. After that, I think libnice has a 5 seconds timeout. But again, I do remember trying to replicate the issue waiting 10 seconds before answering on my SIP softphone and it worked fine, so the cause may be different. I'll check how complicated it is to wait for both offer and answer to be available, hoping it won't reintroduce the infamous bug that almost drove me mad.

**Comment by lminiero** [comment]
Just tried again with this extension in Asterisk:

    exten => 601,1,Wait(40)
    exten => 601,n,Playback(demo-echotest)
    exten => 601,n,Echo
    exten => 601,n,Playback(demo-echodone)
    exten => 601,n,Goto(s,6)

which waits 40 seconds before answering and starting the echo test, and it works as expected for me.

**Comment by saghul** [comment]
> I'll check how complicated it is to wait for both offer and answer to be available, hoping it won't reintroduce the infamous bug that almost drove me mad.

I think this is what we need.

I'll check again, but the problem might surface depending on how many candidates you offer and when. That is, if you trickle all the candidates while the SDP is being processed you may be good, but if you send trickled candidates *after* the offer was processed, ICE will be triggered and it will fail since the browser doesn't know the remote candidates and it's the controlling agent.

**Comment by lminiero** [comment]
Ah but if the problem is when the browser is the controlling agent, then you're experiencing the issue when it's the browser that doesn't answer the call, not when the browser calls and a remote party doesn't anwer... The test I made to replicate the issue has the browser calling, which means it's the controlled agent.

**Comment by lminiero** [comment]
I just tried a test with the browser as the callee, and even in that case it worked as expected, even waiting a lot before answering. I'm afraid I can't replicate the issue you're experiencing :frowning:

**Comment by saghul** [comment]
It's racey, so it's possible you cannot replicate it easily (took me  a while too). Now, let me try to elaborate step by step what I see, because testing alone wasn't enough on my case:

- The browser starts a call (using trickle ICE, and acting as the controlling agent)
- The Janus core parses the SDP
- JANUS_ICE_HANDLE_WEBRTC_PROCESSING_OFFER is set on the handle why processing and cleared afterwards
- The browser sends a trickle message to Janus
- Candidate is parsed using janus_sdp_parse_candidate
- JANUS_ICE_HANDLE_WEBRTC is not yet set because we haven't started the ICE dance
- In sdp.c:524, we start ICE because we have just received a trickled candidate and the offer SDP was already processed

At this point, Janus will feed the ICE agent all the candidates it got from the browser and wait for connectivity checks (Janus is the controlled agent), BUT the browser has no idea about the remote candidates, because Janus hasn't sent the SDP answer back yet. Then ICE will timeout in Janus for any component (ice.c:1120), triggering a hangup event.

The bottom line is that the ICE agent in Janus has knowledge of the local and remote candidates, but the agent in the browser only about the local ones, and since it's the controlling agent, it cannot begin the connectivity checks.

**Comment by lminiero** [comment]
I assume ```JANUS_ICE_HANDLE_WEBRTC``` is ```JANUS_ICE_HANDLE_WEBRTC_START```?

I guess you're getting the same kind of race condition that drove me crazy. What I don't get is that the steps you described should ALWAYS lead to that issue, as that's what always happens when you delay the answer. With the difference being that in all my tests it always works, these are the flags captured via admin API respectively before and after the answer is sent: 

_before_

        "processing-offer": 0,
        "starting": 1,
        "ready": 0,
        "stopped": 0,
        "alert": 0,
        "bundle": 1,
        "rtcp-mux": 1,
        "trickle": 1,
        "all-trickles": 1,
        "trickle-synced": 0,
        "data-channels": 0,
        "has-audio": 1,
        "has-video": 0,
        "plan-b": 0,
        "cleaning": 0

_after_

        "processing-offer": 0,
        "starting": 1,
        "ready": 1,
        "stopped": 0,
        "alert": 0,
        "bundle": 1,
        "rtcp-mux": 1,
        "trickle": 1,
        "all-trickles": 1,
        "trickle-synced": 0,
        "data-channels": 0,
        "has-audio": 1,
        "has-video": 0,
        "plan-b": 0,
        "cleaning": 0

So ```JANUS_ICE_HANDLE_WEBRTC_START``` has indeed been set before the answer was received/handled, but this hasn't resulted in a failure. The only difference in the second dump is that, as expected, the ```JANUS_ICE_HANDLE_WEBRTC_READY``` is set as ICE has succeeded.

I'm wondering if this may be ascribed to different libnice versions? on my new Fedora 22 I now have 0.1.13 installed, while I had 0.1.4 before. Looking at the ICE state in the _before_dump, the reported ICE state was ```connecting``` the whole time, which then turned to ```ready``` when answered. A temporary failure (that is, a failed that is not definitive) should still be reported as ```failed```, as we update the related variable:

https://github.com/meetecho/janus-gateway/blob/master/ice.c#L1118

which seems to suggest this only happens when indeed the libnice stack switch to a ```failed``` state. More recent versions may be more tolerant to make trickling less of an issue.

Anyway, I guess the easiest way to try and fix this is to add a couple of additional flags, e.g., ```JANUS_ICE_HANDLE_WEBRTC_GOT_OFFER``` and ```JANUS_ICE_HANDLE_WEBRTC_GOT_ANSWER```, something that tells us whether or not we have everything we need. We could use those to prevent the definitive failure at line 1120 if we're still lacking an offer or an answer, which would allow ICE to recover when we get them.

**Comment by saghul** [comment]
> Anyway, I guess the easiest way to try and fix this is to add a couple of additional flags, e.g., JANUS_ICE_HANDLE_WEBRTC_GOT_OFFER and JANUS_ICE_HANDLE_WEBRTC_GOT_ANSWER, something that tells us whether or not we have everything we need. We could use those to prevent the definitive failure at line 1120 if we're still lacking an offer or an answer, which would allow ICE to recover when we get them.

Sounds like a good plan to me!

Nevertheless I'll try to see if I can put a new libnice version and re-test, but not relying on libnice specifics would be, well, nice ;-)

**Comment by lminiero** [comment]
Can you check if #322 helps?

**Comment by saghul** [comment]
Will do! Thanks for coooking the patch so quickly :heart_eyes:

**Comment by lminiero** [comment]
This patch only prevents the definitive failed within Janus, meaning it shouldn't send you the hangup before the time, but I'm not sure if libnice will still refuse to do ICE when the browser finally starts working later on (it shouldn't, I remember talking about this with the developers when I wanted to add trickle support to Janus). This means that ICE will still start too soon in Janus, but it shouldn't fail because of this.

**Comment by saghul** [comment]
@lminiero Checked with my colleagues, which were able to reproduce this 100% of the time, and the issue is fixed! :+1:

**Comment by lminiero** [comment]
Yay! I'll merge the pull request then.

**Comment by saghul** [comment]
:cake:
